### ISSUE TITLE
Immutable Collections do not work well with Code Contract Static Analysis

### ISSUE THREAD

**Issue by vossad01** [dotnet/corefx#334]
When modifying a non-null immutable collections we believe the the result should never be null.  However, Code Contract's static analysis is unable to prove that the result is not null generating a lot.  Adding null checks or Contract.Assume statements to pacify the static checker adds unwanted noise to the code.

Would it be possible to add Ensures contracts to Immutable Collections to aid the static analyzer?

### Steps to reproduce:

 1. Create a project
 1. Reference `System.Collections.Immutable (I used 1.1.33-beta from NuGet)
 1. In project properties -> Code Contracts -> Static Checking
    * Check "Perform Static Contract Checking"
    * Set "Warning Level" to "more warnings" (the 3rd from the left)
 1. Add the code below to the project
 1. Build the project
 
``` c#
public class Reproduction
{
	public void ShowContractsIssue()
	{
		IDictionary<string, string> dictionary =
			System.Collections.Immutable.ImmutableDictionary.Create<string, string>()
				.Add("hi", "hello")
				.Add("some", "string");

		this.ActionRequringDictionary(dictionary);
	}

	public void ActionRequringDictionary(IDictionary<string, string> dictionary)
	{
		Contract.Requires(dictionary != null);

		// Other code would usually be here
	}
}
```

### Actual Result:

> Warning
> CodeContracts: requires unproven: dictionary != null. Are you making some assumption on Add that the static checker is unaware of?

### Desired Result:

No warnings because contract was proven

**Comment by Clockwork-Muse** [comment]
@vossad01 - From looking at the code, it appears that the `Ensure()`s are there.  My guess is that the static analysis isn't being exported for external use, though; it's not something the average windows user needs themselves.  Maybe this would result in SDK vs deployed builds.

Of course, there's a whole bunch of other code contract warnings that namespace generates if you run static analysis on it, plus a bunch of missing contract information (there's a bunch of stuff that relies on `Requires.NotNull(...)`, which allows runtime checks, but usually not any accompanying `Contracts.Requires(...)`, so static analysis doesn't happen).  Is anybody planning on fixing those, or should I just start working the warnings?

The `Validation` namespace isn't on here (and I haven't had any luck finding documentation on MSDN or wherever), but it also might be worth looking into annotating some or all of the `Requires` class's methods with `[ContractArgumentValidator]`.

**Comment by ellismg** [comment]
Thanks for bringing this up.  We've discussed our general code contracts strategy and we've decided we're not going to be trying to adopt contracts all over the .NET Core libraries at this time.  See #503 for a further discussion of why.

I believe that the [microsoft/codecontracts](https://github.com/Microsoft/CodeContracts) repository contains a way to produce a set of assemblies with contract annotations in them.  It's possible that the community could contribute there adding annotations for libraries in .NET Core.  However, for the time being, we're not going to try to pull in new contracts or run the contracts rewriter tool.
